### PR TITLE
[Data] Time streaming split overhead

### LINKED ISSUE #42802: [Data] Measure overhead from streaming split
### Description

Measure time spent in `SplitCoordinator` (as part of `streaming_split` path).

### Use case

Additional stats collection for Ray Data performance debugging tool

### PR DESCRIPTION
**Pull request by omatthew98**
## Why are these changes needed?
This adds timing for the overhead created by using the [`streaming_split`](https://docs.ray.io/en/latest/data/api/doc/ray.data.Dataset.streaming_split.html) iterator. There are two primary overheads that are measured:
1. The [OutputSplitter](https://github.com/ray-project/ray/blob/1cc18ea622a6e1899a17f4548aa3734c3656a90f/python/ray/data/_internal/execution/operators/output_splitter.py#L18) operator. This operator is responsible for distributing data to different splits. It runs in the scheduler thread.
2. The [SplitCoordinator actor](https://github.com/ray-project/ray/blob/783da640a20ddbd3b41b893485abf187f0f27223/python/ray/data/_internal/iterator/stream_split_iterator.py#L124). The overhead is mainly from the get and the barrier, which happen in some background threads that handle the remote task requests.

The first overhead is tracked at the operator level in the OutputSplitter operator. The second overhead is tracked at the Dataset level in the SplitCoordinator actor.

## Related issue number
Closes #42802

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/execution/operators/output_splitter.py [hunk 1]
```diff
@@ -233,7 +239,7 @@ def _get_location(self, bundle: RefBundle) -> Optional[NodeIdStr]:
         return bundle.get_cached_location()
 
 
-def _split(bundle: RefBundle, left_size: int) -> (RefBundle, RefBundle):
+def _split(bundle: RefBundle, left_size: int) -> Tuple[RefBundle, RefBundle]:
```
Comment (scottjlee) [inline python/ray/data/_internal/execution/operators/output_splitter.py:242]: are these typehint changes related/needed? same with `_split_meta`, `_split_block`, `_split_single_block`

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 2]
```diff
@@ -235,6 +235,16 @@ def get(
             return block
         except StopIteration:
             return None
+        finally:
+            stats = (
+                self._executor.get_stats()
+                if self._executor
+                else self._base_dataset._plan.stats()
+            )
```
Comment (scottjlee) [inline python/ray/data/_internal/iterator/stream_split_iterator.py:243]: maybe we can rewrite `SplitCoordinator.stats()` to return this instead of `DatasetStatsSummary`, and use the method here

File: python/ray/data/_internal/execution/operators/output_splitter.py (hunk 1 above)
Comment (omatthew98) [reply python/ray/data/_internal/execution/operators/output_splitter.py:242]: They are unrelated but the type `(RefBundle, RefBundle)` isn't a valid type so was just rolling those changes in here since I was touching output_splitter.py. Can remove if it'd be cleaner to separate.

Comment (scottjlee) [reply python/ray/data/_internal/execution/operators/output_splitter.py:242]: SG we can keep it here, thanks

File: python/ray/data/_internal/stats.py [hunk 3]
```diff
@@ -1256,6 +1259,9 @@ def to_string(self) -> str:
                 out += "    * Num blocks unknown location: {}\n".format(
                     self.iter_unknown_location
                 )
+            if self.streaming_split_coord_time.get() != 0:
+                out += "* Streaming split coordinator overhead time: "
```
Comment (scottjlee) [inline python/ray/data/_internal/stats.py:1263]: for formatting, should we omit the * since this is the start of a new section?

**Review by scottjlee — APPROVED** [review]
could you include an example of the new additions to the output string in the PR description?

File: python/ray/data/_internal/stats.py [hunk 4]
```diff
@@ -631,6 +631,9 @@ def __init__(
         self.global_bytes_restored: int = 0
         self.dataset_bytes_spilled: int = 0
 
+        # Streaming split iterator stats
```
Comment (scottjlee) [inline python/ray/data/_internal/stats.py:634]: Let's also note in the comment that this is measured at Dataset level (as opposed to operator level)
```suggestion
        # Streaming split coordinator stats
```

File: python/ray/data/_internal/execution/operators/output_splitter.py [hunk 5]
```diff
@@ -47,6 +48,8 @@ def __init__(
         self._output_queue: deque[RefBundle] = deque()
         # The number of rows output to each output split so far.
         self._num_output: List[int] = [0 for _ in range(n)]
+        # The time of the overhead for the output splitter
```
Comment (scottjlee) [inline python/ray/data/_internal/execution/operators/output_splitter.py:51]: Let's also note in the comment that this is measured at operator level (as opposed to Dataset level)

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 6]
```diff
@@ -204,7 +208,7 @@ def get(
 
         This is intended to be called concurrently from multiple clients.
         """
-
+        start_time = time.perf_counter()
```
Comment (raulchen) [inline python/ray/data/_internal/iterator/stream_split_iterator.py:211]: nit, consider creating a context manager or decorator util to make the code cleaner.

**Review by raulchen — APPROVED** [review]

File: python/ray/data/_internal/iterator/stream_split_iterator.py (hunk 6 above)
Comment (omatthew98) [reply python/ray/data/_internal/iterator/stream_split_iterator.py:211]: I considered that, I think best to leave as is in case the stats object is `None`. I think for my previous PR switching to a context manager caused some integration tests to fail (https://github.com/ray-project/ray/pull/43283). I could use the context manager only when stats is not `None` but not sure if that would end up being cleaner.
